src/librustc_mir/borrow_check.rs

```diff
@@ -1115,13 +1116,238 @@ enum NoMovePathFound {
+        {
+            let mut result = vec![];
+            let mut lvalue = lvalue;
+            loop {
+                result.push(lvalue);
+                match lvalue {
+                    Lvalue::Projection(interior) => {
+                        lvalue = &interior.base;
+                    }
+                    _ => {
+                        result.reverse();
+                        return result;
+                    }
+                }
+            }
+        }
+
+        let borrow_components = lvalue_elements(&borrow.lvalue);
+        let lvalue_components = lvalue_elements(lvalue);
+        debug!("borrow_conflicts_with_lvalue: components {:?} / {:?}",
+               borrow_components, lvalue_components);
+
+        let borrow_components = borrow_components.into_iter()
+            .map(Some).chain(iter::repeat(None));
+        let lvalue_components = lvalue_components.into_iter()
+            .map(Some).chain(iter::repeat(None));
+        for (borrow_c, lvalue_c) in borrow_components.zip(lvalue_components) {
+            // loop invariant: borrow_c is always either equal to lvalue_c or disjoint from it.
+            debug!("borrow_conflicts_with_lvalue: {:?} vs. {:?}", borrow_c, lvalue_c);
+            match (borrow_c, lvalue_c) {
+                (None, _) => {
+                    // the borrow can access our entire lvalue - a conflict.
+                    //
+                    // FIXME: Differs from AST-borrowck; includes drive-by fix
+                    // to #38899. Will probably need back-compat mode flag.
+                    debug!("borrow_conflict_with_lvalue: full borrow, CONFLICT");
+                    return true;
+                }
+                (Some(borrow_c), None) => {
+                    let (base, elem) = match borrow_c {
+                        Lvalue::Projection(box Projection { base, elem }) => (base, elem),
+                        _ => bug!("lvalue has no base?")
+                    };
+                    let base_ty = base.ty(self.mir, self.tcx).to_ty(self.tcx);
+
+                    match (elem, &base_ty.sty, access) {
+                        (_, _, Shallow(Some(ArtificialField::Discriminant))) |
+                        (_, _, Shallow(Some(ArtificialField::ArrayLength))) => {
+                            // The discriminant and array length are like
+                            // additional fields on the type; they do not
+                            // overlap any existing data there. Furthermore,
+                            // they cannot actually be a prefix of any
+                            // borrowed lvalue (at least in MIR as it is
+                            // currently.)
+                            debug!("borrow_conflicts_with_lvalue: implicit field");
+                            return false;
+                        }
+
+                        (ProjectionElem::Deref, _, Shallow(None)) => {
```

Comment:
Comment:

---

Borrowed `*a.b` but accessed `a.b` or some prefix thereof. This is ok for shallow accesses, since they don't deref pointers.